src/Cassandra/Connection.cs

```diff
@@ -600,87 +676,107 @@ public OperationState Send(IRequest request, Action<Exception, AbstractResponse>
+                OperationState state;
+                if (!_writeQueue.TryDequeue(out state))
+                {
+                    //No more items in the write queue
+                    break;
+                }
+                short streamId;
+                if (!_freeOperations.TryPop(out streamId))
                 {
-                    state.Timeout = Configuration.Timer.NewTimeout(() => OnTimeout(delegateState), Configuration.SocketOptions.ReadTimeoutMillis);   
+                    //Queue it up for later.
+                    _writeQueue.Enqueue(state);
+                    //When receiving the next complete message, we can process it.
+                    _logger.Info("Enqueued: {0}, if this message is recurrent consider configuring more connections per host or lower the pressure", _writeQueue.Count);
+                    break;
                 }
+                _logger.Verbose("Sending #{0} for {1}", streamId, state.Request.GetType().Name);
+                _pendingOperations.AddOrUpdate(streamId, state, (k, oldValue) => state);
+                int frameLength;
+                try
+                {
+                    frameLength = state.Request.WriteFrame(streamId, stream);
+                    //Closure state variable
+                    var delegateState = state;
+                    if (Configuration.SocketOptions.ReadTimeoutMillis > 0 && Configuration.Timer != null)
+                    {
+                        state.Timeout = Configuration.Timer.NewTimeout(() => OnTimeout(delegateState), Configuration.SocketOptions.ReadTimeoutMillis);
+                    }
+                }
+                catch (Exception ex)
+                {
+                    //There was an error while serializing or begin sending
+                    _logger.Error(ex);
+                    //The request was not written, clear it from pending operations
+                    RemoveFromPending(streamId);
+                    //Callback with the Exception
+                    state.InvokeCallback(ex);
+                    break;
+                }
+                //We will not use the request any more, stop reference it.
+                state.Request = null;
+                totalLength += frameLength;
             }
-            catch (Exception ex)
+            if (totalLength == 0)
             {
-                //There was an error while serializing or begin sending
-                _logger.Error(ex);
-                //The request was not written, clear it from pending operations
-                RemoveFromPending(streamId);
-                //Callback with the Exception
-                state.InvokeCallback(ex);
+                //nothing to write
+                Interlocked.Exchange(ref _isWriteQueueRuning, 0);
+                return;
             }
+            //Write and close the stream when flushed
+            _tcpSocket.Write(stream.GetBuffer(), (int)stream.Length, () => stream.Close());
```

Comment:
I think that `GetBuffer` should be avoided on a `RecyclableMemoryStream` because it generates large buffer allocations. It would be better to pass the stream to `TcpSocket.Write` and remove the `onBufferFlush` argument.